### DISCUSSION TITLE
why concurrent stress test cpu use 30% increase concurrency can not go up

### DISCUSSION THREAD

**Discussion by demisweety** [libvips/libvips#2408] [Help wanted]
Use the cropping function 50 Concurrent CPU usage rate is 30%, increasing the concurrent CPU usage rate will not go up.

**Comment by jcupitt** [comment]
Hi @demisweety,

It depends on the task -- some just don't have much parallelism. It can also vary with things like speedstep / turboboost etc., where your CPU will clock up and down with load. And it can vary with the libvips cache, which will try to reuse work between threads, when it can.

I used this program for soak testing a while ago. It does 32 resizes in parallel, and repeats 2000 times:

```C
/* Stress-test libvips jpeg read. Read from many threads. 
 *
 * Compile with:
 *
 * 	gcc -g -Wall soak4.c `pkg-config vips --cflags --libs`
 *
 * Run with:
 *
 * 	rm -rf x
 * 	mkdir x
 * 	./a.out ~/pics/k2.jpg 
 */

#include <unistd.h>
#include <vips/vips.h>

/* How many shrinks we run at once.
 */
#define NUM_IN_PARALLEL (32)

/* Number of shrinks we do in total.
 */
#define TOTAL_SHRINKS (NUM_IN_PARALLEL * 2000)

#define OUTPUT_FORMAT ".jpg"

static GThreadPool *pool = NULL;

/* What we run as a thread.
 */
static void 
worker( void *data, void *user_data )
{
	char *file = (char *) data;

	VipsImage *im, *x;
	char output_file[256];

	snprintf( output_file, 256, "x/tmp-%p" OUTPUT_FORMAT, g_thread_self() );

	if( !(im = vips_image_new_from_file( file, 
		"access", VIPS_ACCESS_SEQUENTIAL,
		NULL )) )
		vips_error_exit( NULL );

	if( vips_resize( im, &x, 0.1, NULL ) ) 
		vips_error_exit( NULL );
	g_object_unref( im );
	im = x;

	if( vips_image_write_to_file( im, output_file, NULL ) )
		vips_error_exit( NULL );

	g_object_unref( im );
}

int
main( int argc, char **argv )
{
	int i;

	if( VIPS_INIT( argv[0] ) )
		return( -1 );

	pool = g_thread_pool_new( worker, NULL, NUM_IN_PARALLEL, FALSE, NULL );

	for( i = 0; i < TOTAL_SHRINKS; i++ ) 
		g_thread_pool_push( pool, argv[1], NULL );

	/* Wait for all threads to finish.
	 */
	g_thread_pool_free( pool, FALSE, TRUE );

	sleep( 1 );

	return( 0 );
}
```

I see:

![soak](https://user-images.githubusercontent.com/580843/130228283-0aa2bc07-e5e1-455f-bc5a-28e561d56847.png)

This PC has 32 hardware threads, so it's saturating 25 of them.

**Reply by jcupitt** [reply]
I meant to say, that's on Ubuntu 21.04 and with git master libvips.

Image resize isn't actually very CPU-intensive. If you do something really nasty, it's easy to saturate all hardware threads with a single libvips pipeline. For example:

```
$ vips gaussblur k2.jpg x.jpg 200
```

Hits 3200% CPU use.

You can measure SMP scaling with something like:

```
$ for n in {1..32}; do echo $n; time vips gaussblur k2.jpg x.jpg 200 --vips-concurrency=$n; done
```

Graph that and you'll see it's more or less linear to 16 (the number of cores), then very marginal improvements after that for the hyperthreaded cores.

**Reply by demisweety** [reply]
I use JNI to call (JVips), and it seems that the thread pool is not reused. And simple operations such as cutting to increase concurrency cannot increase CPU usage, nor can it achieve higher performance.I need help!!

**Reply by jcupitt** [reply]
Sorry, I don't see how I can help unless I can see your code, and jvips is nothing to do with me.

Maybe ask the maintainers?
